core/rfb.js

```diff
@@ -420,134 +428,125 @@
                     break;
 
-                case 'connect':
+                case 'connecting':
+                    // A previous connect may asynchronously cause a connection
+                    // so make sure we are still closed.
+                    this._cleanup();
                     this._init_vars();
+
+                    // WebSocket.onopen transitions to the RFB init states
                     this._connect();
-                    // WebSocket.onopen transitions to 'ProtocolVersion'
                     break;
 
-                case 'disconnect':
+                case 'disconnecting':
+                    // WebSocket.onclose transitions to 'disconnected'
+                    this._cleanup();
+
                     this._disconnTimer = setTimeout(function () {
                         this._fail("Disconnect timeout");
                     }.bind(this), this._disconnectTimeout * 1000);
 
                     this._print_stats();
-
-                    // WebSocket.onclose transitions to 'disconnected'
-                    break;
-
-                case 'failed':
-                    if (oldstate === 'disconnected') {
-                        Util.Error("Invalid transition from 'disconnected' to 'failed'");
-                    } else if (oldstate === 'normal') {
-                        Util.Error("Error while connected.");
-                    } else if (oldstate === 'init') {
-                        Util.Error("Error while initializing.");
-                    }
-
-                    // Make sure we transition to disconnected
-                    setTimeout(function () {
-                        this._updateState('disconnected');
-                    }.bind(this), 50);
-
                     break;
 
                 default:
                     // No state change action to take
             }
-
-            if (oldstate === 'failed' && state === 'disconnected') {
-                this._onUpdateState(this, state, oldstate);
-            } else {
-                this._onUpdateState(this, state, oldstate, statusMsg);
-            }
         },
 
         _fail: function (msg) {
-            this._updateState('failed', msg);
+            switch (this._rfb_connection_state) {
+                case 'disconnecting':
+                    Util.Error("Error while disconnecting: " + msg);
+                    break;
+                case 'connected':
+                    Util.Error("Error while connected: " + msg);
+                    break;
+                case 'connecting':
+                    Util.Error("Error while connecting: " + msg);
+                    break;
+                default:
+                    Util.Error("RFB error: " + msg);
+                    break;
+            }
+            this._updateConnectionState('disconnected', msg);
             return false;
         },
 
+        /*
+         * Send a notification to the UI. Accepted optional options are:
+         *    { level: 'warn'|'error' }
```

Comment:
what's the purpose of this being an object?  Why not just an optional "level" argument?